cpp/include/cudf/column/column.hpp

```diff
@@ -0,0 +1,237 @@
+  };
+
+  /**---------------------------------------------------------------------------*
+   * @brief Creates an immutable, non-owning view of the column's data and
+   * children.
+   *
+   * @return column_view The immutable, non-owning view
+   *---------------------------------------------------------------------------**/
+  column_view view() const;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Implicit conversion operator to a `column_view`.
+   *
+   * This allows passing a `column` object directly into a function that
+   * requires a `column_view` and the conversion will happen automatically.
+   *
+   * @return column_view Immutable, non-owning `column_view`
+   *---------------------------------------------------------------------------**/
+  operator column_view() const { return this->view(); };
+
+  /**---------------------------------------------------------------------------*
+   * @brief Creates a mutable, non-owning view of the column's data and
+   * children.
+   *
+   * @note Creating a mutable view of a `column` will invalidate the `column`'s
+   * `null_count()` by setting it to `UNKNOWN_NULL_COUNT`. This will require the
+   * user to either explicitly update the null count with `set_null_count()`,
+   * else, the null count to be recomputed on the next invocation of
+   *`null_count()`.
+   *
+   * @return mutable_column_view The mutable, non-owning view
+   *---------------------------------------------------------------------------**/
+  mutable_column_view mutable_view();
+
+  /**---------------------------------------------------------------------------*
+   * @brief Implicit conversion operator to a `mutable_column_view`.
+   *
+   * This allows pasing a `column` object into a function that accepts a
+   *`mutable_column_view` and the conversion will happen automatically.
+
+   * @note Creating a mutable view of a `column` will invalidate the `column`'s
+   * `null_count()` by setting it to `UKNOWN_NULL_COUNT`. For best performance,
+   * the user should explicitly update the null count with `set_null_count()`.
+   * Otherwise, the null count will be recomputed on the next invocation of
+   * `null_count()`.
+   *
+   * @return mutable_column_view Mutable, non-owning `mutable_column_view`
+   *---------------------------------------------------------------------------**/
+  operator mutable_column_view() { return this->mutable_view(); };
+
+ private:
+  data_type _type{EMPTY};      ///< Logical type of elements in the column
+  cudf::size_type _size{};     ///< The number of elements in the column
+  rmm::device_buffer _data{};  ///< Dense, contiguous, type erased device memory
+                               ///< buffer containing the column elements
+  rmm::device_buffer _null_mask{};  ///< Bitmask used to represent null values.
+                                    ///< May be empty if `null_count() == 0`
+  mutable size_type _null_count{
```

Comment:
Add a `std::string name` member.

Reply:
Based on [the discussion](https://github.com/rapidsai/cudf/issues/1443#issuecomment-485060090) in #1443, I thought the general consensus was to _not_ put a name field in the column directly.  Has that changed or is that considered outside this scope?

Reply:
Yeah, I've been on the fence. After yesterday's meeting there was some more discussion and I was more convinced to add it since enough people wanted it. I don't have strong feelings either way.

So based on the linked discussion, do you think the `name` should go in a "schema" that is part of a table?

Reply:
Yeah, I'd expect the names to be associated with a schema, either stored with the table or one of the results from the I/O loaders.

If the name is associated with the column and libcudf allocates the output columns, how is the name set when libcudf is not loading from a file?  Will it be copied from one of the input columns, or is it up to the caller to set the name after the column is returned since arguably only they care whether a name is set?

Reply:
Do we expect to re-use columns but with different names? For instance, a view of a column which has only been renamed, this the memory can be shared. If so, we'd want to keep the name outside the construct which manages the memory. I can see this happeneing in cudf python, but those values will be copied / marshalled back and forth any way, so the point is probably moot, unless we expect someone to do it in libcudf.

Reply:
I think I'm just going to keep the name out of the column for now. As features are overhauled to use the new structures and the Python column is overhauled, we'll see how the `name` issue shakes out.